File: backend/core/llm.factory.ts
import { ChatGoogleGenerativeAI } from '@langchain/google-genai';
import { ChatOpenAI } from '@langchain/openai';
import { env } from '../config/env.js';
import { agentLogger } from './agent.logger.js';

import { DynamicStructuredTool } from '@langchain/core/tools';

// ─── Singleton State ─────────────────────────────────────────────

let _primaryLlm: ChatOpenAI | null = null; // HuggingFace is now primary
let _fallbackLlm: ChatGoogleGenerativeAI | null = null; // Gemini is now fallback
let _initialized = false;

export const llmInfo = {
	primaryProvider: '',
	primaryModel: '',
	fallbackProvider: '',
	fallbackModel: '',
	mode: 'NONE' as 'dual' | 'primary-only' | 'fallback-only' | 'NONE',
};

function initLLMs() {
	if (_initialized) return;
	_initialized = true;

	// 1. Configure Primary Model (HuggingFace — fast, no rate limit issues)
	if (env.HF_TOKEN) {
		_primaryLlm = new ChatOpenAI({
			model: 'Qwen/Qwen2.5-72B-Instruct',
			apiKey: env.HF_TOKEN,
			configuration: { baseURL: 'https://router.huggingface.co/v1' },
			temperature: 0.3,
			timeout: 30000,
			maxTokens: 1000,
		});
		llmInfo.primaryProvider = 'HuggingFace';
		llmInfo.primaryModel = 'Qwen2.5-72B-Instruct';
		agentLogger.info('LLM', '🟢 Primary LLM configured', {
			provider: 'HuggingFace',
			model: 'Qwen2.5-72B-Instruct',
		});
	} else {
		agentLogger.warn('LLM', '⚠️ HF_TOKEN not set — HuggingFace primary skipped');
	}

	// 2. Configure Fallback Model (Gemini)
	if (env.GEMINI_API_KEY || env.AI_PROVIDER === 'gemini') {
		const apiKey = env.GEMINI_API_KEY || process.env.AI_API_KEY;
		if (apiKey) {
			_fallbackLlm = new ChatGoogleGenerativeAI({
				model: 'gemini-2.5-flash',
				apiKey: apiKey,
				temperature: 0.3,
				maxOutputTokens: 1000,
			});
			llmInfo.fallbackProvider = 'Gemini';
			llmInfo.fallbackModel = 'gemini-2.5-flash-preview-05-20';
			agentLogger.info('LLM', '🟡 Fallback LLM configured', {
				provider: 'Google Gemini',
				model: 'gemini-2.5-flash-preview-05-20',
			});
		}
	}

	// Determine mode
	if (_primaryLlm && _fallbackLlm) {
		llmInfo.mode = 'dual';
		agentLogger.info('LLM', '🚀 Dual-LLM ready: HuggingFace (primary) → Gemini (fallback)');
	} else if (_primaryLlm) {
		llmInfo.mode = 'primary-only';
		agentLogger.info('LLM', '🚀 Single-LLM ready: HuggingFace (no fallback configured)');
	} else if (_fallbackLlm) {
		llmInfo.mode = 'fallback-only';
		agentLogger.warn('LLM', '⚠️ Running on Gemini only (no primary HuggingFace configured)');
	} else {
		llmInfo.mode = 'NONE';
		agentLogger.error('SYSTEM', 'FATAL: No AI Providers configured', null, {
			hint: 'Set HF_TOKEN or GEMINI_API_KEY',
		});
	}
}

// Initialize immediately on module load
initLLMs();

/**
 * Returns the bound primary + fallback LLM instances.
 */
export function getBoundLLMs(tools: DynamicStructuredTool[]) {
	if (llmInfo.mode === 'NONE') {
		throw new Error('No AI Providers configured. Please set HF_TOKEN or GEMINI_API_KEY.');
	}

	return {
		primary: _primaryLlm ? _primaryLlm.bindTools(tools) : null,
		fallback: _fallbackLlm ? _fallbackLlm.bindTools(tools) : null,
		info: llmInfo,
	};
}

File: backend/app/core/llm_factory.py
"""
Dual-LLM factory: HuggingFace Qwen2.5 (primary) + Gemini (fallback).

Same pattern as the Node.js llm.factory.ts — singleton init, tool binding,
and sticky fallback logic.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from langchain_core.tools import BaseTool
from langchain_google_genai import ChatGoogleGenerativeAI
from langchain_openai import ChatOpenAI

from app.config import get_settings
from app.core.logger import agent_logger


@dataclass
class LLMInfo:
    primary_provider: str = ""
    primary_model: str = ""
    fallback_provider: str = ""
    fallback_model: str = ""
    mode: str = "NONE"  # dual | primary-only | fallback-only | NONE


_primary_llm: ChatOpenAI | None = None
_fallback_llm: ChatGoogleGenerativeAI | None = None
_initialized: bool = False
llm_info = LLMInfo()


def _init_llms() -> None:
    global _primary_llm, _fallback_llm, _initialized

    if _initialized:
        return
    _initialized = True

    settings = get_settings()

    # 1. Primary — HuggingFace (OpenAI-compatible endpoint)
    if settings.HF_TOKEN:
        _primary_llm = ChatOpenAI(
            model="Qwen/Qwen2.5-72B-Instruct",
            api_key=settings.HF_TOKEN,
            base_url="https://router.huggingface.co/v1",
            temperature=0.3,
            timeout=30,
            max_tokens=1000,
        )
        llm_info.primary_provider = "HuggingFace"
        llm_info.primary_model = "Qwen2.5-72B-Instruct"
        agent_logger.info("LLM", "🟢 Primary LLM configured", {
            "provider": "HuggingFace", "model": "Qwen2.5-72B-Instruct"
        })
    else:
        agent_logger.warn("LLM", "⚠️ HF_TOKEN not set — HuggingFace primary skipped")

    # 2. Fallback — Gemini
    if settings.GEMINI_API_KEY:
        _fallback_llm = ChatGoogleGenerativeAI(
            model="gemini-2.5-flash",
            google_api_key=settings.GEMINI_API_KEY,
            temperature=0.3,
            max_output_tokens=1000,
        )
        llm_info.fallback_provider = "Gemini"
        llm_info.fallback_model = "gemini-2.5-flash"
        agent_logger.info("LLM", "🟡 Fallback LLM configured", {
            "provider": "Google Gemini", "model": "gemini-2.5-flash"
        })

    # Determine mode
    if _primary_llm and _fallback_llm:
        llm_info.mode = "dual"
        agent_logger.info("LLM", "🚀 Dual-LLM ready: HuggingFace (primary) → Gemini (fallback)")
    elif _primary_llm:
        llm_info.mode = "primary-only"
        agent_logger.info("LLM", "🚀 Single-LLM ready: HuggingFace (no fallback)")
    elif _fallback_llm:
        llm_info.mode = "fallback-only"
        agent_logger.warn("LLM", "⚠️ Running on Gemini only (no primary HuggingFace)")
    else:
        llm_info.mode = "NONE"
        agent_logger.error("SYSTEM", "FATAL: No AI Providers configured", None, {
            "hint": "Set HF_TOKEN or GEMINI_API_KEY"
        })


def get_bound_llms(tools: list[BaseTool]) -> dict[str, Any]:
    """
    Returns the primary + fallback LLMs with tools bound.

    Returns:
        {"primary": bound_llm | None, "fallback": bound_llm | None, "info": LLMInfo}
    """
    _init_llms()

    if llm_info.mode == "NONE":
        raise RuntimeError("No AI Providers configured. Please set HF_TOKEN or GEMINI_API_KEY.")

    return {
        "primary": _primary_llm.bind_tools(tools) if _primary_llm else None,
        "fallback": _fallback_llm.bind_tools(tools) if _fallback_llm else None,
        "info": llm_info,
    }
